chef/cookbooks/barclamp/libraries/barclamp_library.rb

```diff
@@ -417,21 +418,96 @@ def self.size_to_bytes(s)
       end
     end
 
+    class DependsOn
+      class << self
+        def add(dependency)
+          @recipe_depedencies ||= Mash.new
+          @recipe_depedencies.merge!(dependency)
+        end
+
+        def get(recipe)
+          @recipe_depedencies.fetch(recipe, [])
+        end
+      end
+    end
+
     class Config
       class << self
         attr_accessor :node
 
-        def load(group, barclamp, instance = nil)
-          # If no instance is specified, see if this node uses an instance of
-          # this barclamp and use it
+        def last_two_configs(group, barclamp, instance)
+          prev_cfg = load(group, barclamp, instance)
+          instance = infer_instance(group, barclamp, instance)
+          cur_cfg = Chef::Role.load "#{barclamp}-config-#{instance}"
+          return prev_cfg, cur_cfg.default_attributes[barclamp]
+        end
+
+        def loadattr(hash, attrlist)
+          i = 0
+          return hash if attrlist.empty?
+          while hash.respond_to?(:[])
+            unless hash.key?(attrlist[i])
+              Chef::Log.debug("[smart] hash is missing #{attrlist[i]}")
+              return nil
+            end
+            hash = hash[attrlist[i]]
+            i += 1
+            if attrlist.length == i
+              # Chef::Log.debug("[smart] loadattr return #{hash}")
+              return hash
+            end
+          end
+          Chef::Log.debug("[smart] hash does not respond to []")
+          nil
+        end
+
+        def guess_instance(barclamp, instance)
           if instance.nil? && @node[barclamp] && @node[barclamp][:config]
             instance = @node[barclamp][:config][:environment]
           end
 
-          # Accept environments passed as instances
           if instance =~ /^#{barclamp}-config-(.*)/
             instance = $1
           end
+          instance
+        end
+
+        def infer_instance(group, barclamp, instance)
```

Comment:
imho, in software "guess" and "infer" are pretty similar in meaning. I think the names of those two functions should give clear indication on how the instance is "found out".